ENH: Use integer types with fixed width

fbudin69500 opened the pull request:
MetaIO already expects types to have a certain size [1]. (u)int and
(u)long are expected to be encoded on 4 bytes, (u)long long types
are expected to be encoded on 8 bytes. To avoid confusion due to
platform differences, we replace these types that are not visible
or used by the user by fixed width integer types, repectively
(u)int32_t for (u)int and (u)long and (u)unt64_t for (u)long long.
These types are defined in stdint.h

[1] https://github.com/Kitware/MetaIO/blob/master/src/metaTypes.h#L97

aylward commented:
LGTM.   Please verify all ITK tests pass if this update is applied.

blowekamp commented on src/metaTypes.h:
```diff
@@ -35,18 +36,18 @@ namespace METAIO_NAMESPACE {
 typedef char                MET_ASCII_CHAR_TYPE;
 typedef char                MET_CHAR_TYPE;
```
I believe the intention of the "MET_CHAR_TYPE" is to be signed. The "char" type could be either singed or unsigned. I believe that "int8_t" is needed to make it clear how to interpret this type.

fbudin69500 replied:
Good idea. After reading what was said on the ITK mailing list, I agree that (u)int8_t should be used for (unsigned) char to avoid signed/unsigned issues (more specifically to be sure that char is signed since char definition is platform dependent). I will address this issue.

blowekamp commented on src/metaTypes.h:
```diff
@@ -10,6 +10,7 @@
   See the License for more information.
 ============================================================================*/
 #include "localMetaConfiguration.h"
+#include <stdint.h>
```
I believe this header is first available with VS9 SP1. There have been some portability issue with this header. I still carry around a copy in SimpleITK to help with compatibility.

fbudin69500 replied:
@aylward Do we want to support VS2008 (VS9) without SP1 and under?

blowekamp commented on src/metaTypes.h:
```diff
@@ -35,18 +36,18 @@ namespace METAIO_NAMESPACE {
+typedef uint32_t            MET_UINT_TYPE;
+typedef int32_t             MET_LONG_TYPE;
+typedef uint32_t            MET_ULONG_TYPE;
 #if defined(_WIN32) && !defined(__CYGWIN__) && !defined(__MING_W32__)
```
Is this block still needed?

fbudin69500 replied:
Not sure, but since I mostly develop on Linux, I didn't want to break something that was working. A different commit could remove this. This would be cleaner as this might potentially create some issues with very old compilers.

blowekamp commented:
I would go ahead and change the chars to the fixed width to make it clear MetaIO's assumptions. There may be some int8_t->char compatibility issues, and there may be some compatibility with "long" and "long long" for the int64 types too.

blowekamp commented on src/metaTypes.h:
```diff
@@ -35,18 +36,18 @@ namespace METAIO_NAMESPACE {
+typedef int32_t             MET_INT_TYPE;
+typedef uint32_t            MET_UINT_TYPE;
+typedef int32_t             MET_LONG_TYPE;
+typedef uint32_t            MET_ULONG_TYPE;
```
Should these be int64?

fbudin69500 replied:
Looking at [1], it seems to me that (unsigned) long types are defined on 4 bytes (32bits).

[1] https://github.com/Kitware/MetaIO/blob/master/src/metaTypes.h#L63-L97

blowekamp replied:
Well I have not clue how MetaIO works when it thinks MET_ULONG_TYPE is 4 bytes, but it's actually 8!

aylward replied:
Hi,

MET_ULONG_TYPE could be called MET_FOO.   It is unfortunate that the name
doesn't match the construct commonly used now, but it really is ok.

I'd be up for changing the name - but if we do change it, we should
probably change it drastically, so that anyone who was using MET_ULONG_TYPE
will get a compilation error.   For example, we could eliminate
MET_ULONG_TYPE and instead have MET_INT64_TYPE and MET_UINT46_TYPE.

s


On Wed, Sep 14, 2016 at 11:00 AM, Bradley Lowekamp <notifications@github.com
> wrote:

> In src/metaTypes.h
> <https://github.com/Kitware/MetaIO/pull/22#discussion_r78766417>:
>
> > @@ -35,18 +36,18 @@ namespace METAIO_NAMESPACE {
> >  typedef char                MET_ASCII_CHAR_TYPE;
> >  typedef char                MET_CHAR_TYPE;
> >  typedef unsigned char       MET_UCHAR_TYPE;
> > -typedef short               MET_SHORT_TYPE;
> > -typedef unsigned short      MET_USHORT_TYPE;
> > -typedef int                 MET_INT_TYPE;
> > -typedef unsigned int        MET_UINT_TYPE;
> > -typedef long                MET_LONG_TYPE;
> > -typedef unsigned long       MET_ULONG_TYPE;
> > +typedef int16_t             MET_SHORT_TYPE;
> > +typedef uint16_t            MET_USHORT_TYPE;
> > +typedef int32_t             MET_INT_TYPE;
> > +typedef uint32_t            MET_UINT_TYPE;
> > +typedef int32_t             MET_LONG_TYPE;
> > +typedef uint32_t            MET_ULONG_TYPE;
>
> Well I have not clue how MetaIO works when it thinks MET_ULONG_TYPE is 4
> bytes, but it's actually 8!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kitware/MetaIO/pull/22/files/106a72fe8beba785e9ba98c35bb3f2cc32f94618#r78766417>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAREr8J21bOKc2cLuZsW7l5Kbv3Ok-HDks5qqAwDgaJpZM4J8EAc>
> .
>



-- 
====================================================
Stephen R. Aylward, Ph.D.
Senior Director of Operations, North Carolina, Kitware, Inc.
http://www.kitware.com and http://www.aylward.org
(919) 969-6990 x300
====================================================

fbudin69500 replied:
In ITK, the size of MET_LONG_TYPE is compared to the size of long and int [1].
We could indeed rename MET_INT_TYPE to MET_INT32_TYPE and MET_LONG_TYPE to MET_INT64_TYPE as for the NIFTI format [2], and that would be cleaner, but would remove backward compatibility.
[1] https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/IO/Meta/src/itkMetaImageIO.cxx#L212-L219
[2] https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/ThirdParty/NIFTI/src/nifti/niftilib/nifti1.h#L506-L509

blowekamp commented:
This type of change is going to be hard and take time to get it working portably.

It may be worth investing in using CI services to test across platforms and on 32/64-bit compilations?

I believe that ITK contains the needed test suit for MetaIO. I believe you could use ITK for compilation and testing across platforms. The steps would be something like:

1. Checkout ITK (release branch?)
2. do git subtree on MetaIO directory
3. configure with ITK_BUILD_DEFAULT_MODULES:BOOL=OFF ITKIOMeta:BOOL=ON
4. build "ITKIOMeta-all" target
5. run tests with label "ITKIOMeta"

I have been working on scripts to do all but #2 for external ITK modules here[1].

[1] https://github.com/SimpleITK/itkSimpleITKFiltersModule/tree/master/test/ci

fbudin69500 commented:
@blowekamp Because the types are abstracted, both by MetaIO (with the typedef) and in ITK by itkMetaImageIO, I think we are safe making modifications little by little, first in MetaIO, and eventually in ITK. I will start investigating this using the methodology you described. Thanks for your input.

fbudin69500 commented:
@aylward @blowekamp  I added Appveyor support to be able to test MetaIO on Windows. I have updated the tests based on the input I got from @blowekamp to also verify the integration of MetaIO directly in ITK to see if a commit/PR would break the compatibility with ITK. Tests are passing on [TravisCI](https://travis-ci.org/Kitware/MetaIO/builds/165351082) (Linux/MacOS) and on [Appveyor](https://ci.appveyor.com/project/fbudin69500/metaio) (Windows)

blowekamp commented on .travis.yml:
```diff
@@ -1,11 +1,58 @@
+    - PROJ_SRC="${TRAVIS_BUILD_DIR}"
+    - ITK_MODULE_NAME=MetaIO
+    - ITK_BUILD_MODULES="ITKMetaIO;ITKIOMeta"
+    - ITK_TAG=master
```
Could always matrix on this env too, with release if needed :)

fbudin69500 replied:
True, but the matrix would start becoming quite big as I did not find a solution to "summarize" the matrix options, but had to list all individual settings.

blowekamp commented on test/ci/CacheAndUpdateITK.sh:
```diff
@@ -0,0 +1,54 @@
+echo ${ExternalData_OBJECT_STORES}
+mkdir -p "${ExternalData_OBJECT_STORES}"
+
+echo "Using ITK repository: ${ITK_REPOSITORY_REMOTE:=https://github.com/InsightSoftwareConsortium/ITK.git}"
```
Noting that I switch to using the github ITK repo because is was SIGNIFICANTLY faster that the official itk.org one.

fbudin69500 replied:
It would be great to be able to use the same travis_dashboard.cmake and appveyor_dashboard.cmake but it might be difficult as some project might require more environment variables, different build matrix,... That is definitely something to look into it.

fbudin69500 commented:
@blowekamp @aylward I do not have the permissions to this repository to merge this PR. I do not have permission either to test this repository on Appveyor (I ran the test on my own repo).

aylward commented:
I merged.

Perhaps @blowekamp can help with Appveyor.

Thank you!

On Thu, Oct 6, 2016 at 1:12 PM, Francois Budin <notifications@github.com>
wrote:

> @blowekamp <https://github.com/blowekamp> @aylward
> <https://github.com/aylward> I do not have the permissions to this
> repository to merge this PR. I do not have permission either to test this
> repository on Appveyor (I ran the test on my own repo).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kitware/MetaIO/pull/22#issuecomment-252028052>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAREr82I-bAxUsUXx-YZAmO5O_-XF0Nbks5qxSv-gaJpZM4J8EAc>
> .
>



-- 
====================================================
Stephen R. Aylward, Ph.D.
Senior Director of Operations, North Carolina, Kitware, Inc.
http://www.kitware.com and http://www.aylward.org
(919) 969-6990 x300
====================================================

blowekamp commented:
You need admin privileges for a repo to enable a CI system for it, so that the CI can push the needed key into the repo configuration.

I have no privileges here, so unfortunately I am unable to help.


On Oct 6, 2016, at 1:41 PM, Stephen R. Aylward <notifications@github.com<mailto:notifications@github.com>> wrote:

I merged.

Perhaps @blowekamp can help with Appveyor.

Thank you!

On Thu, Oct 6, 2016 at 1:12 PM, Francois Budin <notifications@github.com<mailto:notifications@github.com>>
wrote:

> @blowekamp <https://github.com/blowekamp> @aylward
> <https://github.com/aylward> I do not have the permissions to this
> repository to merge this PR. I do not have permission either to test this
> repository on Appveyor (I ran the test on my own repo).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kitware/MetaIO/pull/22#issuecomment-252028052>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAREr82I-bAxUsUXx-YZAmO5O_-XF0Nbks5qxSv-gaJpZM4J8EAc>
> .
>



--
====================================================
Stephen R. Aylward, Ph.D.
Senior Director of Operations, North Carolina, Kitware, Inc.
http://www.kitware.com and http://www.aylward.org
(919) 969-6990 x300
====================================================

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Kitware/MetaIO/pull/22#issuecomment-252035734>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AATmJRgF9qPBrhZ7tp4Qovja-DUAZh9oks5qxTLXgaJpZM4J8EAc>.

aylward commented:
Both of you are now admins.

I added the CircleCI service and started a build.

On Thu, Oct 6, 2016 at 2:03 PM, Bradley Lowekamp <notifications@github.com>
wrote:

> You need admin privileges for a repo to enable a CI system for it, so that
> the CI can push the needed key into the repo configuration.
>
> I have no privileges here, so unfortunately I am unable to help.
>
>
> On Oct 6, 2016, at 1:41 PM, Stephen R. Aylward <notifications@github.com<
> mailto:notifications@github.com>> wrote:
>
> I merged.
>
> Perhaps @blowekamp can help with Appveyor.
>
> Thank you!
>
> On Thu, Oct 6, 2016 at 1:12 PM, Francois Budin <notifications@github.com<
> mailto:notifications@github.com>>
> wrote:
>
> > @blowekamp <https://github.com/blowekamp> @aylward
> > <https://github.com/aylward> I do not have the permissions to this
> > repository to merge this PR. I do not have permission either to test
> this
> > repository on Appveyor (I ran the test on my own repo).
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Kitware/MetaIO/pull/22#issuecomment-252028052>, or
> mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AAREr82I-bAxUsUXx-
> YZAmO5O_-XF0Nbks5qxSv-gaJpZM4J8EAc>
> > .
> >
>
>
>
> --
> ====================================================
> Stephen R. Aylward, Ph.D.
> Senior Director of Operations, North Carolina, Kitware, Inc.
> http://www.kitware.com and http://www.aylward.org
> (919) 969-6990 x300
> ====================================================
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<https://github.com/
> Kitware/MetaIO/pull/22#issuecomment-252035734>, or mute the thread<
> https://github.com/notifications/unsubscribe-
> auth/AATmJRgF9qPBrhZ7tp4Qovja-DUAZh9oks5qxTLXgaJpZM4J8EAc>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kitware/MetaIO/pull/22#issuecomment-252041652>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARErzalRJ3oIMXrLSwJeoK2A8F7Yv7Fks5qxTfsgaJpZM4J8EAc>
> .
>



-- 
====================================================
Stephen R. Aylward, Ph.D.
Senior Director of Operations, North Carolina, Kitware, Inc.
http://www.kitware.com and http://www.aylward.org
(919) 969-6990 x300
====================================================

fbudin69500 commented:
I don't see a circle.yml file in the repository. Is it in a different branch than master?
PS: Thanks for adding me as an admin. I am looking into adding appveyor CI.
